space_packet_parser/xtcedef.py

```diff
@@ -1243,14 +1300,16 @@ def _get_format_string(self, packet_data: PacketData, parsed_data: dict):
             bits_per_byte = 8
             look_ahead_n_bytes = 0
             while look_ahead_n_bytes <= len(packet_data) - packet_data.pos:
-                look_ahead = packet_data.peek(f'bytes:{look_ahead_n_bytes}')  # Outputs UTF-8 encoded byte string
-                look_ahead = look_ahead.decode('utf-8').encode(self.encoding)  # Force specified encoding
```

Comment:
I don't know why I was decoding in utf8 and then re-encoding in whatever the specified encoding was (e.g. utf-16-be). All we need to do in order to find the termination character is compare literal byte strings, regardless of how they are encoded.

Reply:
I've change the look ahead to look 16 characters at a time rather than one character at a time. This means ~1/16 as many peek operations, and for many strings, we'll catch it in the first peek. And I am splitting the peeked bytes on the termination character now, as you suggested.

Reply:
This is OK, but I guess my more general comment/question was whether we can read the entire string data chunk at once and do this check, rather than the while-loop peeking.

Here is some nice documentation about the various cases:
https://docs.yamcs.org/yamcs-server-manual/mdb/data-types/#string-data-type

I _think_ that my thought is we'd just read the entire "string-buffer" (fixed size of this parameter, or until the end of the packet). That would be our maximum length string. Then the termination character is just used to trim that potential string I think? (i.e I wrote my packet definition to be 8 bytes for this spot, but for this downlink I am only sending `b"HELLO"` and filling 5 of those, with `b"\x00\x00\x00"` that we want to trim off the end and not show to the user.

So, we would always go and get all the bytes for the full string-buffer. Then split on termination character and return the first occurrence. (It also looks like it is acceptable to not terminate the string if the string just ends at the end of the buffer)

The issue here might be that this logic actually needs to be in the `parse_value()` method instead of here. In fact, it might currently be incorrect and setting the position of the data packet to after the termination character rather than after the fixed size field. I don't know enough about all of this, so just putting thoughts down here in case it sparks anything on your end.

Reply:
There are many ways to specify a string length, a termination character is only one. I think if you specify a termination character it has to be there (it can't just end with the end of the buffer). We could peek the entire buffer to find that termination character but that could be 500mb of data if we've read an entire packet file. The basic problem is we have no idea how long the string is. But maybe a compromise could be peeking either the max size of a CCSDS packet from where we currently are, or the full remaining buffer, whichever is shorter. Figure out the string length from that, and send it back.

You're correct that some of this happens in parse_value. That's why this method sends back a skip_bits_after integer. It tells parse_value to skip the termination character after reading the string.

Reply:
In here, we only ever have PacketData being passed, not the entire file. We are always parsing individual packets and the bytes associated with that one piece, so we are helped by that and will only ever read to the end of the packet length.

I just found this: https://github.com/medley56/space_packet_parser/issues/30
and I think your (3) bullet is the issue I think I'm bringing up.

Regardless, this can be considered for follow-up and a future fix and I think we should get this in. I was just bringing it up because it caught my eye here.

Reply:
Ohhh, right! I had forgotten about that. That seems like a nice small ticket to handle next without getting in the way of the bitstring removal and CSV deprecation work. Let's merge this as is, because it does match the current behavior and this kind of logic will still be necessary for `Variable` length defined strings, but this is not its final form.